Dockerfile and instructions

orbitalturtle opened the pull request:
Thanks to @anmode and @tsjk for starting this Dockerfile work in PR #197 . I just made a couple of changes to get it working on my end, including adding the /.teos directory to the teosd group (otherwise I was getting a permissions error when running the container).

- It seems that by only copying over the specific binaries we need (teos-cli and teosd) rather than the full /target folder reduced the build to ~100 MB. 
- Changing from debian to alpine further reduced the build to ~40 MB

For the tutorials, thanks to @sr-gi, who I believe wrote the original Teos python Docker documentation much of which I borrowed in this PR. (Though I removed the mention of creating a volume since we automatically do that in the current Dockerfile. Thoughts?) I've added a section for how to set up a Tor hidden service pointing to the Docker instance.

Next steps:
- One thing that was discussed in PR #197 is to explore how to allow someone to migrate an existing Teos instance to a container. Some changes will be needed to make the existing Tor key compatible and instructions for how to do this need to be added.

(I wrote the Tor Docker instructions after attempting to follow the discussion in #197 . If I got anything wrong, feel free to let me know.)

mariocynicys reviewed (changes requested):
This is missing some docs about `teos-cli`.
The usage should be like so:
```
docker exec -it teos teos-cli --help
```
That's after running `teosd` in a container named `teos` (hence the name after `-it` part)

mariocynicys commented:
I pushed in here by mistake, but how could I even do it :o

mariocynicys commented:
I did some tweaks in [2ad1d4c](https://github.com/talaia-labs/rust-teos/pull/227/commits/2ad1d4c21b47c0de406f402802c925f44d255d4a) to use the default data directory in `/root`, omit making new user to run the tower from, avoid installing bash (not needed actually).

This is the command I run to boot up a dockered teos daemon:
```bash
docker run --network host --name teos -v ./datadir:/root/.teos --rm -it teos
```
Notice the `-v` option to persist the container's datadir in the host FS.

One could use -v with their already populated datadir and have the tower running just fine. but will need to tweak some config settings (shut tor off).

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,78 @@
+- BTC_RPC_PORT=<btc_node_port>
+- BTC_RPC_USER=<btc_rpc_username>
+- BTC_RPC_PASSWORD=<btc_rpc_password>
+- BTC_RPC_CONNECT=<btc_zmq_hostname>
+- BTC_RPC_PORT=<btc_zmq_port>
```
That looks like some data that was not properly removed, it refers to ZMQ which we don't use anymore

sr-gi commented on README.md:
```diff
@@ -23,6 +23,8 @@ Refer to [INSTALL.md](INSTALL.md)
 
 Make sure `bitcoind` is running before running `teosd` (it will fail at startup if it cannot connect to `bitcoind`). [Here](DEPENDENCIES.md#installing-bitcoind) you can find a sample bitcoin.conf.
 
+Please see [Docker instructions](docker/README.md) for instructions on how to set up teos in Docker.
```
:s/teos/`teosd`

orbitalturtle replied:
oh what does :s mean? I assume this means change "teos" to teosd"?

mariocynicys replied:
yup it means `substitute`

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,78 @@
+## Running `teos` in a docker container
```
:s/teos/`teosd`

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,78 @@
+## Running `teos` in a docker container
+A `teos` image can be built from the Dockerfile located in `docker`. You can create the image by running:
```
:s/teos/`teosd`

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,78 @@
+- BTC_RPC_PORT=<btc_zmq_port>
+```
+
+If you are running `teos` and `bitcoind` in the same machine, continue reading for how to create the container based on your OS.
````
:s/teos/`teosd`

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,78 @@
+### `bitcoind` running on the same machine (UNIX)
+The easiest way to run both together in he same machine using UNIX is to set the container to use the host network.
+	
+For example, if both `teos` and `bitcoind` are running on default settings, run
```
:s/teos/`teosd`

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,78 @@
+  -it teos
+```
+
+Notice that you may still need to set your RPC authentication details, since, hopefully, your credentials won't match the `teos` defaults.
````
:s/teos/`teosd`

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,78 @@
+Notice that we also needed to add `API_BIND=0.0.0.0` to bind the API to all interfaces of the container.
+Otherwise it will bind to `localhost` and we won't be able to send requests to the tower from the host.
+
+### How to set up a Tor hidden service pointing to the Docker container:
```
:s/Teosd/`teosd` in the remaining paragraph

sr-gi replied:
Also, I find this last bit hard to follow

orbitalturtle replied:
Ah can you elaborate on which specific part you find hard to follow? I'll try to flesh out the instructions a bit.

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,78 @@
+
+Whereever Tor is running, a user can follow [these instructions](https://community.torproject.org/onion-services/setup/) for setting up a hidden service manually.
+
+For instance, if you're running Teosd on the same machine (similar to the instructions above), you can create a hidden service to hide the IP of the Teosd API (listening on port 9814) by editing your torrc file with the below options (but more likely you'll need to change `127.0.0.1` to the address where your Teos Docker container is running):
```
:s/torrc/`torrc`

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,78 @@
+```
+
+Notice that we also needed to add `API_BIND=0.0.0.0` to bind the API to all interfaces of the container.
+Otherwise it will bind to `localhost` and we won't be able to send requests to the tower from the host.
````
No, localhost inside the container, which maps to something else outside given we are not using the host network in this case

mariocynicys replied:
Gotcha,
Then when to query the tower (e.g. add appointment), what will be the IP to address?

orbitalturtle replied:
If you're accessing it from the host Docker is running on, you can connect to it via localhost

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,78 @@
+For example, if both `teos` and `bitcoind` are running on default settings, run
+    
+```
+docker run --network=host \
+  --name teos \
+  -e BTC_RPC_USER=<rpc username> \
+  -e BTC_RPC_PASSWORD=<rpc password> \
+  -it teos
````
There was a section for that in the original file: https://github.com/talaia-labs/python-teos/blob/master/README.md

orbitalturtle replied:
Not sure if you saw my comment above but since the Dockerfile now automatically creates a volume, I removed these instructions. But happy to swap that out in favor of the volume instructions instead. I have no strong opinions either way.

mariocynicys replied:
Docker creates a new volume each time the container is booted up from the image (unnamed volume) and the volume resides somewhere in `/var/lib/docker`.
With the `-v` it will let the user choose the host mount of the volume so it's reusable with later images.

sr-gi commented on docker/entrypoint.sh:
We are missing some `teosd` options here, are we not?

orbitalturtle replied:
Good call, I'll add the rest of them.

orbitalturtle commented on docker/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+    # Rustfmt is needed to format the grpc stubs generated by tonic.
+    && rustup target add x86_64-unknown-linux-musl\
+    && rustup component add rustfmt\
+    && sed -i -E 's@(^#Crypto.*)@\1\nopenssl = { version = "*", features = [ "vendored"] }@' teos-common/Cargo.toml\
+    && RUSTFLAGS='-C target-feature=+crt-static' cargo build --manifest-path=teos/Cargo.toml --locked --release --target x86_64-unknown-linux-musl
```
Good call

orbitalturtle commented:
Thanks for the reviews! Those changes you made in the Dockerfile @mariocynicys seem good to me. Though I disagree with changing it to run as root - I think we should stick with the principle of least privilege, and I believe that's standard practice https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b

mariocynicys commented:
> Though I disagree with changing it to run as root - I think we should stick with the principle of least privilege, and I believe that's standard practice [medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b](https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b)

Very informative read. I didn't know that a root inside the container is just like a root process in the host :o. Just tried to use one of the protected ports `0-1023` for the tower and it worked just fine with it being root. This is definitely very scary.

mariocynicys commented:
> You can allow others to add commits to your PR if they push to your repo. How did you push to her fork instead of yours, that I don't know lol

I add new remotes to review then checkout the branch locally to push it to my remote instead, forgot to do the last part :joy: :joy:

tsjk commented:
Why does this need `--network=host`?

mariocynicys commented:
> Why does this need --network=host?

running directly on the host network allows us to access bitcoind using `localhost` and allows us to listen for requests on the host port without the need to publish them.

I think we should actually go the Windows & OSX way for linux as well, to:
- Unify how the image is ran over different platforms.
- Limit the root access by not allowing the container to use the host network. (I am kinda convinced now that we can keep running as root user if we have no access to the host network nor the host FS).

sr-gi commented:
> Why does this need `--network=host`?

This is pretty old, but IIRC it was mainly for convenience. It was easier to setup. The MacOS/Windows section was added later in case someone wanted to do it for those platforms, but we expected most people to be running this in UNIX.

orbitalturtle commented:
I finally addressed your comments @mariocynicys and @sr-gi, when you have time for another look. :)

And yes, the tutorial right now assumes you're running Teos in Docker on the same machine as bitcoind. As I eventually get further along with my own Teos Docker setup I'll try to add a guide for someone who might be running bitcoind on a separate raspberry pi, or something, which seems like the more likely real-world scenario.

sr-gi commented on docker/entrypoint.sh:
```diff
@@ -0,0 +1,71 @@
+if [[ ! -z ${DATA_DIR} ]]; then
+    START_COMMAND="$START_COMMAND --datadir $DATA_DIR"
+fi
+
```
Add : # Set boolean flags

The rest of these are booleans, so no parameter is passed. e.g.:

```
START_COMMAND="$START_COMMAND --debug"
```

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,104 @@
+
+### Using `teos-cli`
+
+With this set up, there are two ways you can use `teos-cli`
```
Connecting the `teos-cli` from the host to `teosd` on the guest needs some testing though. I haven't been able to mak it work while testing, even setting the `RPC_BIND` variable (which will be needed AFAICT)

sr-gi reviewed (changes requested):
Mainly nits about structure/naming conventions.

I normally use `teosd` when referring to only the tower backend. `teos` when talking about the project or the system as a whole. I never capitalize this btw, so no Rust-teos, Teos, Teosd, ...

It may be good to test whether you're able to connect a CLN client to the tower running on docker, given the issue I was facing with RPC. This may only apply to macOS/Windows though.

orbitalturtle commented on docker/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+
+RUN apk update && apk upgrade && apk add --update bash
+
+# UID and GID for the TEOSD user
+ENV TEOSD_UID=1001 TEOSD_GID=1001
```
Makes sense

orbitalturtle commented:
@srgi Ahhh yup I now remember the lowercase teos naming conventions from the python teos days :p

I haven't tried connecting using the CLN client yet... But re: connecting to teosd from the host computer with teos-cli. It has been working for me without setting `RPC_BIND`, in linux anyways. Once teosd is running in docker, it should be exposed on localhost from the host. Then I needed to copy the needed tls cert data from the docker /.teos folder to the host's /.teos folder, so that `teos-cli` could connect w/o authentication issues.

What error are you getting? Just not able to connect? When using `docker run`, do you have the `-p 9814:9814` field set?

I'll give it a try on non-linux OSs in VirtualBox some day soon to see if I can come up with anything that'll help

(Also, made those other changes that you requested above)

orbitalturtle commented:
@sr-gi Just confirmed that loading the plugin in CLN and using `lightning-cli registertowe` to connect to the tower running in Docker works for me too

mariocynicys commented on docker/README.md:
```diff
@@ -0,0 +1,104 @@
+You may also want to run docker with a volume, so you can have data persistence in `teosd` databases and keys.
+If so, run:
+
+    docker volume create teos-data
```
The main concern for me here is the configuration file being easily accessible. It's so painful (maybe only for me) having to set the configuration from command line and environment variables and is less maintainable.

orbitalturtle replied:
True, being able to add the config file is much easier than setting the env variables. If I'm understanding right and your goal is mainly to be able to copy the config file into the docker image at /home/.teos, then I don't think the `docker volume create` command is incompatible with that and I don't think we necessarily need to change the mounting instructions

I think the easiest route would be to copy the teos.toml into the working directory before we build the Dockerfile. Then add a line to the dockerfile to copy that file into the container like `COPY teos.toml /home/teosd/.teos`. Or maybe just using `docker cp` after build? https://docs.docker.com/engine/reference/commandline/cp/

I think different devs will favor envs or config file based on their setup. So I don't think we should change our Dockerfile to do this, but instead add instructions for an alternate route instead.

Would this address your concerns there?

mariocynicys replied:
> If I'm understanding right and your goal is mainly to be able to copy the config file into the docker image at /home/.teos

Also onion keys and the DB. But you are right. `docker cp` would do the job I guess.

mariocynicys commented:
Was able to use a non-root user with user-provided data directory: https://github.com/mariocynicys/rust-teos/commit/d2b1f31d5ec9e667a190334e1c9476e3e0192efa

Basically, this change `chown`s the data directory in the container runtime and not the image build time.

tsjk commented on docker/Dockerfile:
```diff
@@ -0,0 +1,50 @@
+RUN apk update && apk upgrade && apk add --update bash
+
+# UID and GID for the teosd user
+ENV TEOS_UID=1001 TEOS_GID=1001
+
+# Copy the teos binaries from the build stage to the new stage
+COPY --from=builder\
+    /tmp/rust-teos/target/x86_64-unknown-linux-musl/release/teosd\
+    /tmp/rust-teos/target/x86_64-unknown-linux-musl/release/teos-cli /usr/local/bin/
+
+# Copy the entrypoint script to the container
+COPY docker/entrypoint.sh /entrypoint.sh
+
+# Set the entrypoint script as executable and add running user
+RUN chmod +x /entrypoint.sh\
+    && addgroup -g ${TEOS_GID} -S teosd\
+    && adduser -S -G teosd -u ${TEOS_UID} teosd
+
+# Expose the default port used by teosd
+EXPOSE 9814/tcp
+
+# Switch user so that we don't run stuff as root
+USER teosd
+
+# Create a volume for the teos data directory
+RUN mkdir /home/teosd/.teos
+RUN chown 1001:1001 /home/teosd/.teos
```
The possibility to sync ids with the world outside is always nice.

mariocynicys replied:
Why `1001` would sync with the host user? Also I am assuming that users will run it without building (the image would be pre-built).

tsjk replied:
Well, certainly one would need to use variables on line 47 as well.
But sure. Whatever. I usually build, and having the option of setting the uid and gid to my ordinary user and not suffer from permission head-aches when sharing data with a container is often appreciated.

tsjk commented on docker/Dockerfile:
```diff
@@ -0,0 +1,50 @@
+EXPOSE 9814/tcp
+
+# Switch user so that we don't run stuff as root
+USER teosd
```
This will not be visible anywhere but inside the container. :)

sr-gi commented:
I need to re-check this on MacOS to see if I'm able to use the CLI/Client from outside

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,104 @@
+For example, if both `teosd` and `bitcoind` are running on default settings, run
+    
+```
+docker run --network=host \
+  --name teos \
````
nit: 

```
docker run \
    --network=host
....
```

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,104 @@
+you can use the special `host.docker.internal` domain.
+
+```
+docker run -p 9814:9814 \
+  --name teos \
````
nit: 

```
docker run \
    -p 9814:9814 \
....
```

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,104 @@
+  -it teos
+```
+
+Notice that we also needed to add `API_BIND=0.0.0.0` to bind the API to all interfaces of the container.
````
It may be nice to add a note on how the same applies to `RPC_BIND`, but I'm still wondering if we can make that work anyway (check main comment)

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,104 @@
+
+`docker exec -it <CONTAINER NAME> sh`
+
+Then you can use the `teos-cli` binary from inside the container as you would use it from your host machine.
```
Not sure this would really work

sr-gi replied:
Can you run `teos-cli` from the host pointing to the container? That should raise an authentication issue, given the credentials of your `.teos` outside the container are different to the ones inside

orbitalturtle replied:
So option 1 is running within the Docker instance so there shouldn't be an authentication issue. But yes, if running from the host machine, then you have to use option 2 (below)

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,104 @@
+
+On the machine where you have Tor running, you can follow [these instructions](https://community.torproject.org/onion-services/setup/) for setting up a hidden service manually.
+
+For instance, if you're running `teosd` in a Docker container on the same machine as where Tor is running, you can create a hidden service from the host machine to hide the IP of the `teosd` API (listening on port 9814). You can do so by editing your `torrc` file on the host machine with the below options:
```
listening on port `9814` for example

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,104 @@
+For instance, if you're running `teosd` in a Docker container on the same machine as where Tor is running, you can create a hidden service from the host machine to hide the IP of the `teosd` API (listening on port 9814). You can do so by editing your `torrc` file on the host machine with the below options:
+
+```
+HiddenServiceDir /var/lib/tor/teosd
````
I think this may depend on the system

orbitalturtle replied:
good point

orbitalturtle replied:
Added a comment about this

sr-gi reviewed (changes requested):
I've been testing this and I was finally able to make it work on MacOS. I think the minimal that is needed is:

```
docker run \
  -p 9814:9814 \
  -p 8814:8814 \
  --name teos \
  --mount source=teos-data,target=/root/.teos \
  -e BTC_RPC_CONNECT=host.docker.internal \
  -e BTC_RPC_USER=<btc_node_hostname> \
  -e BTC_RPC_PASSWORD=<btc_rpc_password>\
  -e API_BIND=0.0.0.0 \
  -e RPC_BIND=0.0.0.0 \
  -it teos
```

This will enable bidirectional communication between the host and the container. However, `teos-cli` may not be usable, giving the credentials may not match.

Also, I've been able to run the Tor bit and interact with a lightning client over Tor

orbitalturtle commented on docker/README.md:
```diff
@@ -0,0 +1,104 @@
+- BTC_RPC_PORT=<btc_node_port>
+- BTC_RPC_USER=<btc_rpc_username>
+- BTC_RPC_PASSWORD=<btc_rpc_password>
+- DATA_DIR=<teos_data_dir>
```
Good call, just removed that option 👍🏻

orbitalturtle commented:
Thanks again for the reviews y'all. This is definitely looking more battle tested. I addressed those comments... I think the top portion of the README had the most changes since I reorganized it to add an option to copy over the config to Docker, upon feedback from @mariocynicys.

@sr-gi Awesome, glad to hear you got it working on MacOS. I updated the instructions with that new command, if that seems like a good enough fix. 

Hoping it's close! 🤞🏻

sr-gi commented on docker/README.md:
````diff
@@ -0,0 +1,131 @@
+For instance, if you're running `teosd` in a Docker container on the same machine as where Tor is running, you can create a hidden service from the host machine to hide the IP of the `teosd` API (listening on port 9814 for example). If you're using Linux, you can do so by editing your `torrc` file on the host machine with the below option:
+
+```
+HiddenServiceDir /var/lib/tor/teosd # On Linux, this is the directory this data is typically stored. This directory can be whereever you want, but if you'd like it to be stored with other Tor data, check what the Tor data directory is on your particular operating system.
````
I'll keep this comment simpler:

Path for Linux. This may change depending on your OS.

sr-gi commented on docker/README.md:
```diff
@@ -0,0 +1,131 @@
+  --network=host
+  --name teos \
+  --mount source=teos-data,target=/home/teos/.teos \
+  -e BTC_RPC_CONNECT=host.docker.internal \
```
I don't think this should be needed for Linux, given we are using --nerwork=host. Can you confirm?

orbitalturtle replied:
Yes you're correct 👍🏻

sr-gi reviewed (approved):
LGTM

mariocynicys commented on docker/entrypoint.sh:
```diff
@@ -0,0 +1,67 @@
+    START_COMMAND="$START_COMMAND --btcrpcport $BTC_RPC_PORT"
+fi
+
+if [[ ! -z ${DEBUG} ]]; then
+    START_COMMAND="$START_COMMAND --debug"
+fi
+
+if [[ ! -z ${DEPS_DEBUG} ]]; then
+    START_COMMAND="$START_COMMAND --depsdebug"
+fi
+
+if [[ ! -z ${OVERWRITE_KEY} ]]; then
+    START_COMMAND="$START_COMMAND --overwritekey"
+fi
+
+if [[ ! -z ${FORCE_UPDATE} ]]; then
+    START_COMMAND="$START_COMMAND --forceupdate"
+fi
```
These four don't need to be set to some string, thus we should use `! -z ${VARIABLE+x}` which checks if the variable is set even if it is an empty string.

orbitalturtle replied:
Hmmm I worry slightly about this because if someone is used to setting the variables as true or false in the config variable might be confused when they set one of the values above to "false" but this logic considers this to be "true"

But I definitely see what you're saying. What do you think about doing something like this: setting the above as boolean strings (since in shell script booleans are... confusing). We write a comment noting that to turn them on they must be set the string to "true". And then we change the above logic to just `if [ "$DEBUG" = true ]; then` or something along those lines.

mariocynicys replied:
Agreed!

orbitalturtle replied:
Good catch. I made those updates, let me know what you think

orbitalturtle commented on docker/Dockerfile:
```diff
@@ -0,0 +1,50 @@
+# Use a new stage with a smaller base image to reduce image size
+FROM alpine:latest
+
+RUN apk update && apk upgrade && apk add --update bash
```
Ah good catch

orbitalturtle commented on docker/README.md:
```diff
@@ -0,0 +1,130 @@
+
+2) Config file:
+
+If you'd like to use a `teos.toml` file instead, one option is to edit the Dockerfile with this line right above the ENTRYPOINT:
+
+	COPY teos.toml /home/teos/.teos
```
I didn't realize this before, but `docker cp` won't actually work here because it can only be used once the container is running, but we need to copy over this data `before` it's running and we hit the endpoint. 

Can you elaborate on what you mean by the second point?

mariocynicys replied:
> I didn't realize this before, but docker cp won't actually work here because it can only be used once the container is running, but we need to copy over this data before it's running and we hit the endpoint.

The way I did it (but think it's tedious tbh) is spin up a dummy alpine container with the volume just to copy the data over to the volume with `docker cp`, and then reuse the volume with `teos` image.

> Can you elaborate on what you mean by the second point?

When a volume is attached, any image data inside the mount point disappears, I guess what docker does is just replace the directory from the image with the volume content. So the `COPY` command provides some data at build time, which if you run the container without a volume would persist as you would expect, but if a volume is attached, the mount directory would be identical to what's inside the volume, effectively ignoring the build-time `COPY` command.

orbitalturtle replied:
Oh got it, interesting. Well one option in that case would be to add a VOLUME command above the COPY command in the Dockerfile to create the volume before rather than after

Personally, I don't think we don't need to get super deep in these instructions into... what if a user does this, what will happen? It's nice that we have some instructions to give users an outline, but I suspect most people using this Dockerfile will have some knowledge of Docker and will know how to tweak it for their specific needs

mariocynicys replied:
Fine, let's remove the config file instructions and maybe just mention that such a method exists and leave it up to the user how to do so.

tsjk commented on docker/Dockerfile:
```diff
@@ -0,0 +1,50 @@
+    && rustup target add x86_64-unknown-linux-musl\
+    && rustup component add rustfmt\
+    # Cross compile with musl as the target, so teosd can run on alpine
+    && RUSTFLAGS='-C target-feature=+crt-static' cargo build --manifest-path=teos/Cargo.toml --locked --release --target x86_64-unknown-linux-musl
```
Compiling statically does make sense as we want everything in the two binaries. If it works without, then maybe the devs have made it default. I think they need to give feedback here.

tsjk replied:
But maybe it's just the default with musl. C.f. https://rust-lang.github.io/rfcs/1721-crt-static.html

orbitalturtle replied:
Hmmm got it, I'd be down to keep it in there just to err on the side of caution

mariocynicys reviewed (approved):
LGTM :)

orbitalturtle commented:
@mariocynicys Yay! Made that last remaining change.

sr-gi commented:
Re-ACK [c3c7e99](https://github.com/talaia-labs/rust-teos/pull/227/commits/c3c7e9939e28f0005d5fd71236fed44b6aaa898d)

mariocynicys commented:
ACK [c3c7e99](https://github.com/talaia-labs/rust-teos/pull/227/commits/c3c7e9939e28f0005d5fd71236fed44b6aaa898d)

sr-gi commented:
Needs rebase

orbitalturtle commented:
@sr-gi @mariocynicys just rebased
